### PR TITLE
Add unit tests covering core packaging script functionality

### PR DESCRIPTION
**Pull request by raramakr**
Includes tests for all functions in the packaging scripts, input validation, metadata extraction, and packaging logic. These tests strengthen confidence in the build pipeline and help prevent subtle breakages when the script is modified.

### CODE REVIEW

File: build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md
Comment (ScottTodd) [inline build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md:1]: There is little in this file that I see as worth committing to the repository. Much of the file summarizes what already exists in code and therefore is at risk of going stale (e.g. "54 tests", "96% code coverage")

File: build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+python3 -m unittest discover -v
+```
+
+### Run Tests for a Specific Module
+
+To run tests for `packaging_utils.py`:
+
+```bash
+python3 -m unittest test_linux_packaging_utils -v
+```
```
Comment (ScottTodd) [inline build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md:27]: These are generic instructions for how to run Python unit tests. They would apply equally to any directory containing `_test.py` files.

File: build_tools/packaging/tests/test_linux_packaging_utils.py
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:1]: Match project style, name tests `*_test.py` and not `test_*.py`

See also https://docs.pytest.org/en/stable/explanation/goodpractices.html#conventions-for-python-test-discovery (both naming conventions work, but we should be consistent)

File: build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+## Requirements
+
+The tests require:
+
+- Python 3.6+
+- Standard library modules: `unittest`, `tempfile`, `pathlib`, `json`, `os`, `shutil`
+- The modules being tested: `packaging_utils.py` and `build_package.py`
+
+No additional packages like pytest are required.
```
Comment (ScottTodd) [inline build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md:185]: These requirements are not accurate or useful.

* We typically support Python 3.11+  at the project level. No need to say that the tests for one subdirectory support Python 3.6+
* No need to comment on usage of standard library modules or that tests need the system under test to be run
* Depending on any packages already in https://github.com/ROCm/TheRock/blob/main/requirements-test.txt is fine

What _would_ be useful would be a "how to run TheRock's Python unit tests" general guide that covers all of the `build_tools/` directory equally. I'm exploring adding code coverage measurements across that directory and I could put docs for that in the same file.

File: build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+## Contributing
+
+When adding new functions to `packaging_utils.py` or `build_package.py`:
+
+1. Create a corresponding test class in the appropriate test file
+1. Test both success and failure cases
+1. Test edge cases and error conditions
+1. Use mocks to avoid external dependencies
+1. Run all tests to ensure no regressions
```
Comment (ScottTodd) [inline build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md:195]: We should run these tests on CI too, then there is no need to document things like "when changing to code, ensure tests are still passing" - CI will just enforce that.

We currently run this code in our CI workflows: https://github.com/ROCm/TheRock/blob/13adb6ec50e762fde8eeaa152a896bb35a08fd18/.github/workflows/build_portable_linux_artifacts.yml#L106-L108

See example logs: https://github.com/ROCm/TheRock/actions/runs/21163169347/job/60861823710?pr=2992#step:8:1

We can extend that to also include this `build_tools/packaging/tests/` path.

---

Note that https://github.com/ROCm/TheRock/blob/main/build_tools/packaging/tests/promote_from_rc_to_final_test.py is designed to be run manually, but it is already properly excluded:
```
(.venv) λ pytest build_tools\packaging\tests/
================================================= test session starts ================================================= platform win32 -- Python 3.13.5, pytest-8.4.1, pluggy-1.6.0
rootdir: D:\projects\TheRock
plugins: hydra-core-1.3.2, check-2.5.3
collected 0 items

================================================ no tests ran in 0.25s ================================================
```

File: build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+## Test Features
+
+### Mocking
+
+Tests use Python's `unittest.mock` module to:
+
+- Mock file system operations
+- Mock subprocess calls (dpkg-buildpackage, rpmbuild)
+- Mock external dependencies
+- Isolate units under test
```
Comment (ScottTodd) [inline build_tools/packaging/tests/LINUX_PACKAGING_TEST_README.md:130]: This sort of "testing practices" documentation could go in the style guide here: https://github.com/ROCm/TheRock/blob/main/docs/development/style_guides/python_style_guide.md#testing-standards

It is _not_ worth including in a readme file this deeply nested into a subfolder.

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+        """Test that print_function_name prints the calling function name."""
+
+        def caller_function():
+            packaging_utils.print_function_name()
+
+        caller_function()
+        mock_print.assert_called_once()
+        # Check that the printed string contains "In function:"
+        call_args = mock_print.call_args[0][0]
+        self.assertIn("In function:", call_args)
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:35]: This is a useless unit test. The function it is testing contains a single `print()` call: https://github.com/ROCm/TheRock/blob/13adb6ec50e762fde8eeaa152a896bb35a08fd18/build_tools/packaging/linux/packaging_utils.py#L44-L51

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 6]
```diff
@@ -0,0 +1,863 @@
...
+        new_callable=mock_open,
+        read_data='[{"Package": "test-pkg"}]',
+    )
+    def test_read_package_json_file_success(
+        self,
+        mock_file,
+    ):
+        """Test successful reading of package.json file."""
+        result = packaging_utils.read_package_json_file()
+        self.assertEqual(result, [{"Package": "test-pkg"}])
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:59]: Mocking the path open defaults the purpose of testing this function. The function loads https://github.com/ROCm/TheRock/blob/main/build_tools/packaging/linux/package.json, so the test should load the real file.

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 7]
```diff
@@ -0,0 +1,863 @@
...
+            "TRUE",
+            "True",
+            "YES",
+        ]
+        for val in true_values:
+            pkg_info = {"TestKey": val}
+            self.assertTrue(
+                packaging_utils.is_key_defined(pkg_info, "TestKey"),
+                f"Failed for value: {val}",
+            )
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:86]: More useless test code here. This is just a change detector test, matching the `true_values` list in the test against the list in the `is_key_defined()` function.

A more useful test would
1. Load the real file (instead of creating a test-only `pkg_info` dict)
2. Check that a known-valid key is defined in the package info
3. Check that a known-missing key is NOT defined in the package info

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+
+    def test_postinstall_true(self):
+        """Test when Postinstall is enabled."""
+        pkg_info = {"Postinstall": "true"}
+        self.assertTrue(packaging_utils.is_postinstallscripts_available(pkg_info))
+
+    def test_postinstall_false(self):
+        """Test when Postinstall is disabled."""
+        pkg_info = {"Postinstall": "false"}
+        self.assertFalse(packaging_utils.is_postinstallscripts_available(pkg_info))
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:142]: More trivial test code here, not worth including

https://github.com/ROCm/TheRock/blob/13adb6ec50e762fde8eeaa152a896bb35a08fd18/build_tools/packaging/linux/packaging_utils.py#L116-L127

The docstring says "is enabled *__for a package__*" (emphasis mine), yet I see only one package.json. Are there going to be multiple package.json files in the future, or is this `is_postinstallscripts_available()` function just always going to return `True`?

Similar comments for the functions that just call `is_key_defined()` with no extra logic. That's a lot of boilerplate (in the file itself, docstrings, and now in tests) for just `return True`.

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 9]
```diff
@@ -0,0 +1,863 @@
...
+        ]
+        result = packaging_utils.get_package_list()
+        self.assertEqual(result, ["pkg1", "pkg3"])
+
+    @patch("packaging_utils.read_package_json_file")
+    def test_get_package_list_all_enabled(self, mock_read):
+        """Test get_package_list when all packages are enabled."""
+        mock_read.return_value = [{"Package": "pkg1"}, {"Package": "pkg2"}]
+        result = packaging_utils.get_package_list()
+        self.assertEqual(result, ["pkg1", "pkg2"])
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:269]: I think it would be more useful to test without mocks here. Load the real file and check that some packages that are expected to be there are indeed returned.

If loading the file is slow and there are lots of test cases that would be doing it again and again, it could be done once as a test-suite setup step.

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 10]
```diff
@@ -0,0 +1,863 @@
...
+        mock_print.assert_called_once()
+        self.assertIn("Removed directory", mock_print.call_args[0][0])
+
+    @patch("builtins.print")
+    def test_remove_nonexistent_dir(self, mock_print):
+        """Test removing a non-existent directory."""
+        test_dir = Path(self.temp_dir) / "nonexistent"
+        packaging_utils.remove_dir(test_dir)
+        mock_print.assert_called_once()
+        self.assertIn("does not exist", mock_print.call_args[0][0])
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:302]: Not a useful test. The code just runs `shutil.rmtree()`. We don't need to test the standard library.

https://github.com/ROCm/TheRock/blob/13adb6ec50e762fde8eeaa152a896bb35a08fd18/build_tools/packaging/linux/packaging_utils.py#L291-L306

File: build_tools/packaging/tests/test_linux_packaging_utils.py [hunk 11]
```diff
@@ -0,0 +1,863 @@
...
+
+
+class TestVersionToStr(unittest.TestCase):
+    """Test cases for version_to_str function."""
+
+    def test_version_three_parts(self):
+        """Test version with three parts."""
+        self.assertEqual(packaging_utils.version_to_str("7.1.0"), "70100")
+        self.assertEqual(packaging_utils.version_to_str("10.1.0"), "100100")
+        self.assertEqual(packaging_utils.version_to_str("1.2.3"), "10203")
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_packaging_utils.py:312]: These tests ARE actually useful, thanks.

(the code being tested has nontrivial application logic that is important for the packaging process)

File: build_tools/packaging/tests/test_linux_build_package.py [hunk 12]
```diff
@@ -0,0 +1,1174 @@
...
+        self.source_dir = Path(self.temp_dir) / "source"
+        self.dest_dir = Path(self.temp_dir) / "dest"
+        self.source_dir.mkdir()
+
+    def tearDown(self):
+        """Clean up temporary directories."""
+        if os.path.exists(self.temp_dir):
+            shutil.rmtree(self.temp_dir)
+
+    @patch("builtins.print")
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_build_package.py:38]: Why are these test cases patching `print`? Logging during tests is fine (and even useful). We could also switch to using `logging` with more precise control over log levels.

File: build_tools/packaging/tests/test_linux_build_package.py [hunk 13]
```diff
@@ -0,0 +1,1174 @@
...
+            shutil.rmtree(self.temp_dir)
+
+    @patch("build_package.append_version_suffix")
+    @patch("build_package.is_meta_package")
+    @patch("build_package.convert_to_versiondependency")
+    @patch("build_package.update_package_name")
+    @patch("builtins.print")
+    def test_generate_control_file_versioned(
+        self, mock_print, mock_update, mock_convert, mock_is_meta, mock_append
+    ):
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_build_package.py:232]: Too much use of `patch` here. Can't tell what the test is even trying to do.

File: build_tools/packaging/tests/test_linux_build_package.py [hunk 14]
```diff
@@ -0,0 +1,1174 @@
...
+            "7.1.0",
+        ]
+
+        build_package.main(argv)
+
+        mock_run.assert_called_once()
+        args = mock_run.call_args[0][0]
+        self.assertEqual(args.rocm_version, "7.1.0")
+        self.assertEqual(args.pkg_type, "deb")
+        self.assertEqual(args.target, "gfx900")
```
Comment (ScottTodd) [inline build_tools/packaging/tests/test_linux_build_package.py:688]: We don't need to test the standard library (argparse).

**Review by ScottTodd — CHANGES_REQUESTED** [review]
This PR pursues test coverage at the expense of so, so many engineering practices. I'm not sure where to even start with review until the tests actually consider what is worth testing.

I would focus testing on:

* Application logic (e.g. version parsing, file format validation, package construction)
* rpm/deb spec compliance
* debuggable integration tests (like https://github.com/ROCm/TheRock/blob/main/build_tools/packaging/tests/promote_from_rc_to_final_test.py)

**Review by stellaraccident — CHANGES_REQUESTED** [review]
I didn't review this PR in detail, but I recognize the use of unconstrained AI in test generation when I see it. I note many testing styles that you actively have to steer against vs accept (unwarranted use of mocks, change detectors, std library testers, fragile print-based testing (with mocks), etc.

When I use AI to generate tests, I will note that it is almost always in conjunction with code changes to optimize for testability. If you haven't done that work up front or as part of a refactor done in conjunction with test coverage enhancement, the AI throws up its hands and does the only thing it can do. It is the author's job to architect both the code under test and the test coverage to be useful.

**Comment by raramakr** [comment]
> I didn't review this PR in detail, but I recognize the use of unconstrained AI in test generation when I see it. I note many testing styles that you actively have to steer against vs accept (unwarranted use of mocks, change detectors, std library testers, fragile print-based testing (with mocks), etc.
> 
> When I use AI to generate tests, I will note that it is almost always in conjunction with code changes to optimize for testability. If you haven't done that work up front or as part of a refactor done in conjunction with test coverage enhancement, the AI throws up its hands and does the only thing it can do. It is the author's job to architect both the code under test and the test coverage to be useful.

Ack , Will  make the changes to make the test code more meaningful

**Comment by raramakr** [comment]
This is already addressed in https://github.com/ROCm/TheRock/commit/282cbc2cab1a03eb8bce2b10a172ae1af193abd5
